File: dsa/7.linked_list/203_Remove_Linked_List_Elements.py
import typing as t
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def removeElements(self, head: t.Optional[ListNode], val: int) -> t.Optional[ListNode]:
        prev_node: ListNode = None
        pointer_node: ListNode = head
        while pointer_node:
            next_node = pointer_node.next
            if pointer_node.val == val:
                if prev_node:
                    prev_node.next = next_node
                else:
                    pointer_node.next = None
                    head = next_node
            else:
                prev_node = pointer_node
            pointer_node = next_node
        

        return head

File: dsa/7+8.linked_list/203_Remove_Linked_List_Elements.java
package linked_list;

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
class Solution {
    public ListNode removeElements(ListNode head, int val) {
        if (head == null){
            return null;
        }

        ListNode curNode = head;
        ListNode prevNode = null;

        while (curNode != null){
            ListNode surNode = curNode.next;
            if (curNode.val == val){
                if (prevNode != null){
                    prevNode.next = surNode;
                    curNode = prevNode;
                }
                else{
                    head = surNode;
                    curNode = null;
                }
            }
            prevNode = curNode;
            curNode = surNode;
        }
        return head;
    }
}
